[Order Creation] Use store currency for all new orders

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14357
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

On multicurrency sites, the user’s default currency is used for order creation when none is specified in the API request.

This may differ from the site’s default currency. A user can set their default currency at `/my-account/edit-account/`.

Previously we didn’t specify a currency, so if the merchant’s default was different from the site’s, they weren’t be able to use in-person payments.

Additionally, order creation previously showed that they were creating the order in the site’s default currency, but when it was created it actually used the user’s, so it appeared to switch over, which was confusing.

To work around this in a world where the app doesn’t really support multicurrency, we set the site’s default currency code in the initial order creation request.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Using a multi-currency site (can be set up in WooPayments), log in to your user account on the site (`https://siteurl/my-account/edit-account/`.)

At the bottom, you can select a currency to be your user's default – change this to be something other than the site default currency.

Log in to the app using the same user.

Create an order with products or a custom amount – observe that it uses the site currency.

Repeat with POS.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

It's worth quickly testing order editing too, as it uses the same code, though editing orders which are in another currency is disabled so this is simplified quite a bit.

When I first tested this in POS, we didn't use the same currency for the item list as for the order, though the store currency was used for the order.

Note that the logic for determining whether there are any changes in the order creation screen (this decides whether it can be closed without a warning) has changed slightly, so check that cancelling an order creation screen still works as intended as well.

I improved this warning in split view; previously it would only show the dialog if the order had synced changes, now it shows the warning for unsynced changes too.

I've tested on an iPad Air running iOS 17.7.2, and an iPhone 15 running iOS 18.1.1

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

My test store uses GBP as it's currency, while my user has the preference set for EUR.

### Before

https://github.com/user-attachments/assets/4e720d02-d24f-4d62-b162-c7237da01430

### After

https://github.com/user-attachments/assets/5c15e6a2-2db0-48f3-8243-e90128a2fce7


https://github.com/user-attachments/assets/366b9c3a-7a8c-4a36-b73b-cbbfb910436c



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
14357 Use store currency for all new orders

On multicurrency sites, the user’s default currency is used for order creation when none is specified in the API request.

This may differ from the site’s default currency. A user can set their default currency at `/my-account/edit-account/`.

Previously we didn’t specify a currency, so if the merchant’s default was different from the site’s, they weren’t be able to use in-person payments.

Additionally, order creation previously showed that they were creating the order in the site’s default currency, but when it was created it actually used the user’s, so it appeared to switch over, which was confusing.

To work around this in a world where the app doesn’t really support multicurrency, we set the site’s default currency code in the initial order creation request.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
14357 Add site currency default to 21.5 release

joshheald committed:
14357 Use initial order to determine changes

The order synchronizer previously used a static let of an empty order to make it’s new order.

The same static property was used to check whether the order had been edited or not.

Because Order has dates in it, the static property would have had the dates from the initial call to it, as long as the app was running.

When we added the currency parameter, this made it clunkier to use the static property approach, and it made little sense to have out of date dates anyway.

This commit updates the order sychronizer to be responsible for keeping track of whether the order has changed or not. It’s still done using equality, but we compare to the order that we initially created, rather than the static property.

joshheald committed:
14357 fix order store tests

joshheald committed:
14537 Ensure the currency field is passed for POS

joshheald committed:
14357 Improve cancelation warning in split view

joshheald committed:
Merge branch 'trunk' into issue/14357-use-site-currency-for-order-creation

staskus commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -63,11 +63,11 @@ final class EditableOrderViewModel: ObservableObject {
-            return orderSynchronizer.order != OrderFactory.emptyNewOrder
-        case .editing(let initialOrder):
-            return orderSynchronizer.order != initialOrder
+        if selectionSyncApproach == .onRecalculateButtonTap {
```
There are no issues with this logic but it revealed an issue with `canBeDismissed: Bool` when testing.

```
    var canBeDismissed: Bool {
        switch flow {
        case .creation: // Creation can be dismissed when there aren't changes pending to commit.
            return !hasChanges
        case .editing: // Editing can always be dismissed because changes are committed instantly.
            return true
        }
    }
```

When we edit on an iPad, then if we make changes in a product picker and click dismiss, we don't show an alert.

I think it should be:

```
    var canBeDismissed: Bool {
        switch flow {
        case .creation: // Creation can be dismissed when there aren't changes pending to commit.
            return !hasChanges
        case .editing: // Editing can always be dismissed on iPhone because changes are committed instantly.
            return !(selectionSyncApproach == .onRecalculateButtonTap && hasChanges)
        }
    }
```

joshheald replied:
Agreed... however I've just had a go at fixing that, and opens a can of worms which I don't want to dive in to right now.

1. If you change it as suggested, discarding the changes during editing deletes the order.
2. If you prevent the deletion, making an edit, tapping cancel, then continuing with the edit instead of discarding it breaks the `Recalculate` button (and I have no idea why...)

To get most of the fix out, I'm going to log this as a separate issue... but it'll be pretty low priority, I think.

staskus replied:
Got it, thanks for checking that!

staskus reviewed (approved):
👏 Great work! Thanks for addressing this.

It's great that we can combine explicitly setting site currency on the order with not allowing to edit order in non-site currency. 

That was a bit of headache how to combine both the creation and editing and it's great to have a consistent solution now that prefers site currency on mobile payments 👍
